src/app-admin.js

```diff
@@ -0,0 +1,491 @@
+    return null;
+  }
+
+  if (!(await isAdmin(session.user_id))) {
+    sendErrorResponse(res, 401, 'unauthorized', 'permission-not-found');
+    return null;
+  }
+
+  return session;
+};
+
+const validateUserInput = (res, user) => {
+  if (!user.userName) {
+    sendErrorResponse(
+      res,
+      400,
+      'user-cant-be-empty',
+      'Username cannot be empty',
+    );
+    return false;
+  }
+
+  if (!user.role) {
+    sendErrorResponse(res, 400, 'role-cant-be-empty', 'Role cannot be empty');
+    return false;
+  }
+
+  const { id: roleId } =
+    getAccountDb().first('SELECT id FROM roles WHERE roles.id = ?', [
+      user.role,
+    ]) || {};
+
+  if (!roleId) {
+    sendErrorResponse(
+      res,
+      400,
+      'role-does-not-exists',
+      'Selected role does not exists',
+    );
+    return false;
+  }
+
+  return true;
+};
+
+app.get('/masterCreated/', (req, res) => {
+  const { cnt } =
+    getAccountDb().first(
+      `SELECT count(*) as cnt
+     FROM users
+     WHERE users.user_name <> '' and users.master = 1`,
+    ) || {};
+
+  res.json(cnt > 0);
+});
+
+app.get('/users/', (req, res) => {
+  const session = validateUser(req, res);
+  if (!session) return;
+
+  const users = getAccountDb().all(
+    `SELECT users.id, user_name as userName, display_name as displayName, enabled, ifnull(master,0) as master, roles.id as role 
+     FROM users
+     JOIN user_roles ON user_roles.user_id = users.id
+     JOIN roles ON roles.id = user_roles.role_id
+     WHERE users.user_name <> ''`,
+  );
+
+  res.json(
+    users.map((u) => ({
+      ...u,
+      master: u.master === 1,
+      enabled: u.enabled === 1,
+    })),
+  );
+});
+
+app.post('/users', async (req, res) => {
+  const user = await getAdminSessionFromRequest(req, res);
+  if (!user) return;
+
+  const newUser = req.body;
+  const { id: userIdInDb } = getUserByUsername(newUser.userName);
+
+  if (!validateUserInput(res, newUser)) return;
+  if (userIdInDb) {
+    sendErrorResponse(
+      res,
+      400,
+      'user-already-exists',
+      `User ${newUser.userName} already exists`,
+    );
+    return;
+  }
+
+  const userId = uuid.v4();
+  let displayName = newUser.displayName || null;
+  let enabled = newUser.enabled ? 1 : 0;
+
+  getAccountDb().mutate(
+    'INSERT INTO users (id, user_name, display_name, enabled, master) VALUES (?, ?, ?, ?, 0)',
+    [userId, newUser.userName, displayName, enabled],
+  );
+
+  getAccountDb().mutate(
+    'INSERT INTO user_roles (user_id, role_id) VALUES (?, ?)',
+    [userId, newUser.role],
+  );
+
+  res.status(200).send({ status: 'ok', data: { id: userId } });
+});
+
+app.patch('/users', async (req, res) => {
+  const user = await getAdminSessionFromRequest(req, res);
+  if (!user) return;
+
+  const userToUpdate = req.body;
+  const { id: userIdInDb } =
+    getAccountDb().first('SELECT id FROM users WHERE id = ?', [
+      userToUpdate.id,
+    ]) || {};
+
+  if (!validateUserInput(res, userToUpdate)) return;
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I think validation should be moved up to line 159 (before querying the DB).

Reply:
moved